src/shims/windows/sync.rs

```diff
@@ -130,4 +224,198 @@ pub trait EvalContextExt<'mir, 'tcx: 'mir>: crate::MiriEvalContextExt<'mir, 'tcx
+
+        let addr = ptr.addr().bytes();
+
+        let compare = this.deref_operand(compare_op)?;
+        let size = this.read_scalar(size_op)?.to_machine_usize(this)?;
+
+        let timeout_ms = this.read_scalar(timeout_op)?.to_u32()?;
+
+        let timeout_time = if timeout_ms == this.eval_windows("c", "INFINITE")?.to_u32()? {
+            None
+        } else {
+            this.check_no_isolation("`WaitOnAddress` with non-infinite timeout")?;
+
+            let duration = Duration::from_millis(timeout_ms as u64);
+
+            Some(Time::Monotonic(Instant::now().checked_add(duration).unwrap()))
+        };
+
+        let Some(layout) = this.wait_on_address_layout(size) else {
+            let invalid_param = this.eval_windows("c", "ERROR_INVALID_PARAMETER")?;
+            this.set_last_error(invalid_param)?;
+            this.write_scalar(Scalar::from_i32(0), dest)?;
+            return Ok(())
+        };
+
+        let futex_val = this.read_futex_val(ptr_op, layout)?;
+
+        let undesired = this
+            .read_scalar(&compare.into())?
+            .check_init()?
+            .to_uint(layout.size)?
+            .try_into()
+            .expect("layout should be only u8, u16, u32 or u64 from wait_on_address_layout");
+
+        if futex_val == undesired {
+            this.block_thread(thread);
+            this.futex_wait(addr, thread, u32::MAX, Some((undesired, size)));
+
+            if let Some(timeout_time) = timeout_time {
+                let dest = *dest;
+
+                this.register_timeout_callback(
+                    thread,
+                    timeout_time,
+                    Box::new(move |this| {
+                        this.unblock_thread(thread);
+                        this.futex_remove_waiter(addr, thread);
+                        let error_timeout = this.eval_windows("c", "ERROR_TIMEOUT")?;
+                        this.set_last_error(error_timeout)?;
+                        this.write_scalar(Scalar::from_i32(0), &dest)?;
+                        Ok(())
+                    }),
+                )
+            }
+        }
+
+        this.write_scalar(Scalar::from_i32(1), dest)?;
+
+        Ok(())
+    }
+
+    fn WakeByAddressSingle(&mut self, ptr_op: &OpTy<'tcx, Tag>) -> InterpResult<'tcx> {
+        let this = self.eval_context_mut();
+
+        let ptr = this.read_pointer(ptr_op)?;
+
+        if let Some(waiter) = this.futex_wake(ptr.addr().bytes(), u32::MAX) {
+            let (undesired, size) =
+                waiter.undesired.expect("undesired value should always exist on windows");
+
+            let layout = this
+                .wait_on_address_layout(size)
+                .expect("should have been checked in WaitOnAddress");
+
+            let futex_val = this.read_futex_val(ptr_op, layout)?;
+
+            if futex_val != undesired {
+                this.unblock_thread(waiter.thread);
+                this.unregister_timeout_callback_if_exists(waiter.thread);
+            }
```

Comment:
So this is why we now expose this `undesired` thing... is this protection against spurious wakeups? Please explain in comments.
https://devblogs.microsoft.com/oldnewthing/20160823-00/?p=94145 sounds like spurious wakeups are allowed, so I am not sure why you are protecting against them here.

What happens if they are not equal? You dequeued the thread, so nobody else is going to wake it up, but you are not waking it up either. It will just sleep forever?